### PR TITLE
Allow composite functions and TF Asymmetry in DoublePulseFit

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
**Description of work.**

This updates DoublePulseFit to allow composite functions to be used. It also updates CalculateMuonAsymmetry to allow double pulse fits to be performed in tf asymmetry mode.

**To test:**

The following script performs a DoublePulseFit and the equivalent normal fit against some artificial data. Check that the results match and that the script runs without issue
```
delta = 0.33
x = np.linspace(0.,15.,100)
x_offset = np.linspace(delta/2, 15. + delta/2, 100)
x_offset_neg = np.linspace(-delta/2, 15. - delta/2, 100)

testFunction = GausOsc(Frequency = 1.5, A=0.22)
y1 = testFunction(x_offset_neg)
y2 = testFunction(x_offset)
y = y1/2+y2/2 + 3.0
ws = CreateWorkspace(x,y)

convolution =  FunctionFactory.createCompositeFunction('Convolution')
innerFunction = FunctionFactory.createInitialized('name=GausOsc,A=0.2,Sigma=0.2,Frequency=1,Phi=0; name=FlatBackground, A0=5.0')
deltaFunctions = FunctionFactory.createInitialized('(name=DeltaFunction,Height=0.5,Centre={},ties=(Height=0.5,Centre={});name=DeltaFunction,Height=0.5,Centre={},ties=(Height=0.5,Centre={}))'.format(-delta/2, -delta/2, delta/2, delta/2))
convolution.setAttributeValue('FixResolution', False)
convolution.add(innerFunction)
convolution.add(deltaFunctions)

innerFunctionSingle = FunctionFactory.createInitialized('name=GausOsc,A=0.2,Sigma=0.2,Frequency=1,Phi=0; name=FlatBackground, A0=5.0')

DoublePulseFit(Function=innerFunctionSingle, InputWorkspace=ws, CreateOutput = True, PulseOffset = delta, StartX=0.0, EndX=15.0, Output='DoublePulseFit', MaxIterations=100)
Fit(Function=convolution, InputWorkspace=ws, CreateOutput = True, StartX=0.0, EndX=15.0, Output='Fit', MaxIterations=100)
```

This script performs a double pulse Fit in CalculateMuonAsymmetry
```
delta = 0.33
x = np.linspace(0.,15.,100)
x_offset = np.linspace(delta/2, 15. + delta/2, 100)
x_offset_neg = np.linspace(-delta/2, 15. - delta/2, 100)

testFunction = GausOsc(Frequency = 1.5, A=0.22)
y1 = testFunction(x_offset_neg)
y2 = testFunction(x_offset)
N0 = 6.38
y = N0 * (1 + y1/2+y2/2)
y_norm = y1/2+y2/2
unnormalised_workspace = CreateWorkspace(x,y)
ws_to_normalise = CreateWorkspace(x,y)
ws_correctly_normalised = CreateWorkspace(x,y_norm)
AddSampleLog(Workspace='ws_to_normalise', LogName="analysis_asymmetry_norm", LogText="1")

innerFunction = FunctionFactory.createInitialized('name=GausOsc,A=0.20,Sigma=0.2,Frequency=1.0,Phi=0')
tf_function = ConvertFitFunctionForMuonTFAsymmetry(InputFunction=innerFunction, WorkspaceList=['ws_to_normalise'])

CalculateMuonAsymmetry(MaxIterations=100, EnableDoublePulse=True, PulseOffset=delta, UnNormalizedWorkspaceList='unnormalised_workspace', ReNormalizedWorkspaceList='ws_to_normalise',
                               OutputFitWorkspace='DoublePulseFit', StartX=0, InputFunction=str(tf_function), Minimizer='Levenberg-Marquardt')
double_parameter_workspace = AnalysisDataService.retrieve('DoublePulseFit_Parameters')
values_column = double_parameter_workspace.column(1)
# Check that normalised data is correct
 result, message = CompareWorkspaces('ws_to_normalise', 'ws_correctly_normalised', Tolerance=1e-10)
        
```

Also check that the workflows work in the Muon Analysis GUI by:
  * Performing a fit of a GausOsc and a flat background in double pulse mode.
  * Performing a Double pulse fit in tf asymmetry mode.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Muon/src/CalculateMuonAsymmetry.cpp [hunk 1]
```diff
@@ -93,6 +93,13 @@ void CalculateMuonAsymmetry::init() {
   declareProperty(std::make_unique<API::FunctionProperty>(
                       "OutputFunction", Kernel::Direction::Output),
                   "The fitting function after fit.");
+  declareProperty(
+      "EnableDoublePulse", false,
+      "Controls whether to perform a double pulse or single pulse fit.");
+  declareProperty("PulseOffset", 0.0, "The time offset between the two pulses");
+  declareProperty("FirstPulseWeight", 0.5,
+                  "Weighting of first pulse."
```
Comment (AnthonyLim23) [inline Framework/Muon/src/CalculateMuonAsymmetry.cpp:101]:
```suggestion
                  "Weighting of first pulse (w_1)."
```

File: Framework/Muon/src/CalculateMuonAsymmetry.cpp [hunk 2]
```diff
@@ -93,6 +93,13 @@ void CalculateMuonAsymmetry::init() {
...
+                  "The second pulse weighting is set as 1 - this.");
```
Comment (AnthonyLim23) [inline Framework/Muon/src/CalculateMuonAsymmetry.cpp:102]:
```suggestion
                  "The second pulse weighting (w_2) is set as w_2 = 1 - w_1.");
```

File: docs/source/release/v5.1.0/muon.rst [hunk 3]
```diff
@@ -41,6 +41,8 @@ Improvements
 - Improved the maintainability of the fitting code in the Muon Analysis interface by removing redundant and duplicated code.
   These changes will make the addition of new functionality in the future easier.
 - The plot guess option within the fitting tab will now update when a parameter is changed.
+- Updated :ref:`DoublePulseFit <algm-DoublePulseFit>` to allow composite function input.
```
Comment (AnthonyLim23) [inline docs/source/release/v5.1.0/muon.rst:44]: is this one supposed to be here? Can we also have a line saying TF asym and double pulse is now available in the GUI?

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py [hunk 4]
```diff
@@ -573,19 +575,39 @@ def get_parameters_for_single_tf_fit(self, workspace):
...
             'ReNormalizedWorkspaceList': workspaces,
             'UnNormalizedWorkspaceList': self.context.group_pair_context.get_unormalisised_workspace_list(workspaces),
             'OutputFitWorkspace': fit_workspaces,
             'StartX': self.fitting_options["startX"],
             'EndX': self.fitting_options["endX"],
-            'Minimizer': self.fitting_options["minimiser"]
+            'Minimizer': self.fitting_options["minimiser"],
         }
+        if ('DoublePulseEnabled' in self.context.gui_context and self.context.gui_context['DoublePulseEnabled']):
+            offset = self.context.gui_context['DoublePulseTime']
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py:603]: lets make this a function and we can call it here and in the above

File: docs/source/release/v5.1.0/muon.rst (hunk 3 above)
Comment (Mantid-Matthew) [reply docs/source/release/v5.1.0/muon.rst:44]: Yes this change is also in this PR. I have added a line abut the GUI.

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py (hunk 4 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py:603]: I meant the contents of the if

**Comment by AnthonyLim23** [comment]
can we add a python example to the TF Asymm docs?

**Review by AnthonyLim23 — DISMISSED** [review]
Looks good to me :shipit:

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py [hunk 5]
```diff
@@ -177,7 +179,7 @@ def do_single_tf_fit(self, parameter_dict):
 
     def do_single_fit_and_return_workspace_parameters_and_fit_function(
             self, parameters_dict):
-        if 'DoublePulseEnabled' in self.context.gui_context and self.context.gui_context['DoublePulseEnabled']:
+        if self.double_pulse_enanbled():
```
Comment (martyngigg) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py:182]: Sorry to be really picky. There is a typo here that my eyes now can't unsee...

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py (hunk 5 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py:182]: thats a good spot. The if statement would not check correctly with that typo

**Review by martyngigg — APPROVED** [review]

**Pull request merged**